src/sys/sgx/tcp/stream.rs

```diff
@@ -0,0 +1,488 @@
+                write_buffer: WriteBuffer::new(User::<[u8]>::uninitialized(WRITE_BUFFER_SIZE)),
+                #[cfg(not(compiler_has_send_sgx_types))]
+                write_buffer: MakeSend::new(WriteBuffer::new(User::<[u8]>::uninitialized(WRITE_BUFFER_SIZE))),
+                write_state: State::New(()),
+                #[cfg(compiler_has_send_sgx_types)]
+                read_buf: Some(User::<[u8]>::uninitialized(READ_BUFFER_SIZE)),
+                #[cfg(not(compiler_has_send_sgx_types))]
+                read_buf: Some(MakeSend::new(User::<[u8]>::uninitialized(READ_BUFFER_SIZE))),
+                read_state: State::New(()),
+                registration: None,
+                provider: None,
+            }))),
+        }
+    }
+
+    pub(super) fn from_std(stream: net::TcpStream) -> (Self, SocketAddr) {
+        let peer_addr = stream.peer_addr().unwrap_or_else(|_| ([0; 4], 0).into());
+        let stream = TcpStream::new(State::Ready(stream));
+        (stream, peer_addr)
+    }
+
+    pub fn connect(addr: SocketAddr) -> io::Result<TcpStream> {
+        Ok(TcpStream::new(State::New(addr.to_string())))
+    }
+
+    pub fn connect_str(addr: &str) -> io::Result<TcpStream> {
+        Ok(TcpStream::new(State::New(addr.to_owned())))
+    }
+
+    pub fn peer_addr(&self) -> io::Result<SocketAddr> {
+        self.inner()
+            .connect_state
+            .as_ready()
+            .ok_or_else(|| would_block())
+            .and_then(|stream| stream.peer_addr())
+    }
+
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.inner()
+            .connect_state
+            .as_ready()
+            .ok_or_else(|| would_block())
+            .and_then(|stream| stream.local_addr())
+    }
+
+    pub fn shutdown(&self, _how: Shutdown) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
+    }
+
+    pub fn set_nodelay(&self, _nodelay: bool) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
+    }
+
+    pub fn nodelay(&self) -> io::Result<bool> {
+        Ok(false) // ineffective in SGX
+    }
+
+    pub fn set_ttl(&self, _ttl: u32) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
+    }
+
+    pub fn ttl(&self) -> io::Result<u32> {
+        Ok(DEFAULT_FAKE_TTL) // ineffective in SGX
+    }
+
+    pub fn take_error(&self) -> io::Result<Option<io::Error>> {
+        let mut inner = self.inner();
+        if let Some(err) = inner.connect_state.take_error(State::Error(io::ErrorKind::Other.into())) {
+            return Ok(Some(err));
+        }
+        if let Some(err) = inner.read_state.take_error(State::New(())) {
+            return Ok(Some(err));
+        }
+        if let Some(err) = inner.write_state.take_error(State::New(())) {
+            return Ok(Some(err));
+        }
+        Ok(None)
+    }
+
+    pub fn peek(&self, _buf: &mut [u8]) -> io::Result<usize> {
+        Ok(0) // undocumented current behavior in std::net::TcpStream for SGX target.
```

Comment:
This probably needs to be documented. Are there plans to securely implement `peek()` at some point?

Reply:
I think it will only be implemented when there's a strong need for it. There are no plans I'm aware of.